students/Zach Cooper/lesson04/basic_operations_refactored.py

```diff
@@ -0,0 +1,150 @@
+""" Basic Operations"""
+import logging
+import csv
+import peewee as pw
+from customers_model import Customer
+import create_customer as cc
+
+
+LOG_FORMAT = "%(asctime)s %(filename)s:%(lineno)-3d %(levelname)s %(message)s"
+FORMATTER = logging.Formatter(LOG_FORMAT)
+
+FILE_HANDLER = logging.FileHandler('db.log')
+FILE_HANDLER.setFormatter(FORMATTER)
+
+LOGGER = logging.getLogger(__name__)
+LOGGER.addHandler(FILE_HANDLER)
+
+LOGGER.info("Basic operations defined for the customer database.")
+
+
+def add_customer(customer_id, first_name, last_name, home_address,
+                 phone_number, email_address, status, credit_limit):
+    """
+        Returns new customer with **kwargs info
+    """
+    try:
+        credit_limit = float(credit_limit)
+    except ValueError as err:
+        logging.error(err)
+        raise
+    try:
+        with Customer.DATABASE.transaction():
+            new_customer = Customer.create(
+            customer_id=customer_id,
+            first_name=first_name,
+            last_name=last_name,
+            home_address=home_address,
+            phone_number=phone_number,
+            email_address=email_address,
+            status=status,
+            credit_limit=credit_limit)
+            new_customer.save()
+
+            LOGGER.info('Customer database add successful')
+            LOGGER.info('%s %s has been created', Customer.first_name,
+                Customer.last_name)
+    except pw.IntegrityError:
+        LOGGER.error("Customer id %s, already exitsts in database. Error adding %s %S",
+                    customer_id, first_name, last_name)
+    raise pw.IntegrityError
+
+def upload_csv(filename):
+    """
+        Adds data for customer from csv file
+    """
+    with open(filename, newline="", encoding="ISO-8859-1") as csvfile:
+        all_customers = csv.reader(csvfile)
+        header = next(all_customer, None) 
+
+        with Customer.DATABASE.atomic(): # Populates the databse with info
+            Customer.insert_many(all_customers, headers).execute
+
+
+def search_customer(customer_id):
+    """
+        Search for customer name in database
+        :return: Customer info if Ture, empty dict if None
+        :param customer_id: customer_id to search for
+    """
+    find_customer_dict = {}
+    LOGGER.info("Search for customer name in database...")
+    try:
+        find_customer = Customer.get(Customer.customer_id == customer_id)
+        if find_customer:
+
+            find_customer_dict = {"first_name": find_customer.first_name,
+                                  "last_name": find_customer.last_name,
+                                  "email_address": find_customer.email_address,
+                                  "phone_number": find_customer.phone_number}
```

Comment:
Here is a place for a comprehension, but you have to know about getattr. Getattr gets a attribute from an object, so find_customer.first_name is equivalent to getattr(find_customer, 'first_name'), which is easier to substitute a variable in.
```
my_attrs = ['first_name', 'last_name', 'email_address', 'phone_number']
find_customer_dict = {cust_attr: getattr(find_customer, cust_attr) for cust_attr in my_attrs}
```